alerting:labels-annotations elaboration (#86663)
grafana/grafana: docs/sources/alerting/fundamentals/alert-rules/annotation-label.md

# Labels and annotations

> Labels and annotations contain information about an alert. Labels are used to differentiate an alert from all other alerts, while annotations are used to add additional information to an existing alert.

Reviewer: what's wrong with this sentence? For me it offers a quick definition and distinction up front.

Reviewer: Labels and annotations contain information about an alert and can be used to optimize and customize your alert rules.

Author: nothing wrong actually!
restored that paragraph

Reviewer: haha:)

> When creating alert rules, you can leverage labels and annotations to optimize and customize your alerts. **Labels** serve as unique identifiers of an alert. And, beyond their labeling function, they also offer practical functionalities such as searchability, silencing, routing, and templating.

Reviewer: ```suggestion
Labels are unique identifiers of an alert. You can also use them to search, silence, route, and template.
```

Reviewer: Can you use labels to silence? i would probably add "route notifications" and i don't think you use labels to template but rather you can template labels? (the other way around?)

Author: >Can you use labels to silence?
yep in fact labels (.i,e, label matcher) is what you use to select which alerts to silence.

> I don't think you use labels to template but rather you can template labels? (the other way around
true. That's the way to express it. Changing that.Good catch!

> On the other hand, **annotations** provide contextual metadata at specific time points, enriching alerts with additional information. They also are key/value pairs, but unlike labels, annotations focus on offering insights of the envents or incidents, providing descriptions or summaries to help you understand the data being alerted on by using a combination of text/template.

Reviewer: ```suggestion
Annotations provide contextual metadata at specific time points.  They give insights into events or incidents by providing descriptions or summaries to help you understand the data being alerted on. 
```

Reviewer: what do you mean by provide metadata at specific time points?
I think key value/pairs etc needs to be explained later on - it seems out of context in a quick overview?
these descriptions or summaries appear in the notification messages, right? so maybe "annotations provide additional information in notification messages to help you understand the data that is being alerted on".

Reviewer: i wrote this for the in product help, maybe we could just use this: 
Annotations add metadata to provide more information on the alert in your alert notification messages. For example, add a Summary annotation to tell you which value caused the alert to fire or which server it happened on. Annotations can contain a combination of text and template code.

Author: > what do you mean by provide metadata at specific time points?
i meant to say that annotation includes data about the data (.i.e metrics) but i will keep it simpler and remove that part 

I made a combination of your suggestions. let me know!

Labels and annotations contain information about an alert. Labels are used to differentiate an alert from all other alerts, while annotations are used to add additional information to an existing alert.

> When creating alert rules, you can leverage labels and annotations to optimize and customize your alerts. **Labels** serve as unique identifiers of an alert. And, beyond their labeling function, they also offer practical functionalities such as searchability, silencing, routing notifications, and they can be templated.

Reviewer: ```suggestion
When creating alert rules, you can leverage labels and annotations to optimize and customize your alerts. 

Labels are unique identifiers of an alert rule. They can also be used for searching, silencing, and routing notifications. 
```

Reviewer: You can also template labels and annotations.

Reviewer: i don't see this change yet..maybe you missed it in the midst of all the others haha

Author: ![image](https://github.com/grafana/grafana/assets/45235678/81fd3914-6732-48a6-8497-0ae88d461bfc)
somehow it is not reflected in here but i did push it :)

> On the other hand, **annotations** provide additional information in notification messages to help you understand the data that is being alerted on. Annotations also give more information on the alert in your alert notification messages. For example, you can add a Summary annotation to tell you which value caused the alert to fire or which server it happened on. Annotations can contain a combination of text and template code.

Reviewer: ```suggestion
Annotations provide additional information in alert notification messages to help you understand the data that is being alerted on. For example, you can add a **Summary** annotation to tell you which value caused the alert rule to fire or which server it happened on. Annotations can contain a combination of text and template code.
```

Therefore, labels and annotations can help you to manage your alert rules, and to receive more flexible and informative alert notifications by applying the use of templates.

> 

Reviewer: I would delete this; you've said that all above

When creating alert rules, you can also template labels and annotations to optimize and customize your alerts.

## Labels

**Labels** serve as unique identifiers of an alert. And, beyond their labeling function, they also offer practical functionalities such as searchability, silencing, routing notifications, and they can be templated.

> Labels contain information that identifies an alert. An example of a label might be `server=server1` or `team=backend`. Each alert can have more than one label, and the complete set of labels for an alert is called its label set. It is this label set that identifies the alert.

Reviewer: ```suggestion
Examples of labels are `server=server1` or `team=backend`. Each alert rule can have more than one label and the complete set of labels for an alert rule is called its label set. It is this label set that identifies the alert.
```

> For example, an alert might have the label set `{alertname="High CPU usage",server="server1"}` while another alert might have the label set `{alertname="High CPU usage",server="server2"}`. These are two separate alerts because although their `alertname` labels are the same, their `server` labels are different.

Reviewer: ```suggestion
For example, an alert rule might have the label set `{alertname="High CPU usage",server="server1"}` while another alert rule might have the label set `{alertname="High CPU usage",server="server2"}`. These are two separate alert rules because although their `alertname` labels are the same, their `server` labels are different.
```

> An alert's label set can contain three types of labels:

Reviewer: i corrected these to alert rule - but i'm not sure anymore. is it alert rule, or alert instance? or...?

Author: alert rule is correct 😃

> - The data source query. For example, if you were monitoring temperature readings and each time series for these readings would have a sensor_id, and a location label. Those labels are used to provide additional context or dimensions to the metric data, helping to differentiate between different time series.

Reviewer: ```suggestion
- Data source query labels. For example, if you are monitoring temperature readings and each time series for these readings has a sensor_id, and a location label. These labels are used to provide additional context or dimensions to the metric data, helping to differentiate between different time series.
```

> - Some are automatically added by Grafana (i.e. alertname and grafana_folder). These are Grafana reserved labels.

Reviewer: ```suggestion
- Labels that are automatically added by Grafana (i.e. alertname and grafana_folder). These are Grafana reserved labels.
```

> - Others you can define yourself to help filter data in your alert rules. They act as a unique identifier of an alert. In addition, labels allow searching, silencing, or routing alerts. 

Reviewer: ```suggestion
- Labels that you define yourself to help filter data in your alert rules. 
```

> And, you can also template them. For example in your alert, you could add a label that uses templating to create more dynamic and customizable alerting. E.g. `environment` `=` `{{ your text/template }}`.

Reviewer: ```suggestion
You can also template labels. For example in your alert rule, you could add a label that uses templating to create more dynamic and customizable alerting. E.g. `environment` `=` `{{ your text/template }}`.
```

{{<admonition type="note">}}

> Ensure the label set for an alert does not have two or more labels with the same name. If a label has the same name as a label from the datasource then it will replace that label. However, should a label have the same name as a reserved label then the label will be omitted from the alert.

Reviewer: ```suggestion
Ensure the label set for an alert does not have two or more labels with the same name. If a label has the same name as a label from the data source then it will replace that label. However, should a label have the same name as a reserved label then the label will be omitted from the alert.
```
